src/coreclr/vm/amd64/AsmHelpers.asm

```diff
@@ -558,17 +559,307 @@ NESTED_ENTRY InterpreterStub, _TEXT
 
         PROLOG_WITH_TRANSITION_BLOCK
 
-        ;
-        ; call ExecuteInterpretedMethod
-        ;
-        lea             rcx, [rsp + __PWTB_TransitionBlock]     ; pTransitionBlock*
-        mov             rdx, METHODDESC_REGISTER
-        call            ExecuteInterpretedMethod
+        __InterpreterStubArgumentRegistersOffset = __PWTB_ArgumentRegisters
+        ; IR bytecode address
+        mov             rbx, METHODDESC_REGISTER
+
+        INLINE_GETTHREAD r13; thrashes rax and r11
+
+        ; Load the InterpMethod pointer from the IR bytecode
+        mov             rax, qword ptr [METHODDESC_REGISTER]
+        mov             rax, qword ptr [rax + OFFSETOF__InterpMethod__pCallStub]
+        and             rax, rax
+        jne             HaveCallStub
+        ; The call stub was not generated yet, so we need to do it now
+        mov             rcx, qword ptr [METHODDESC_REGISTER]
+        ; Interpreter-TODO: consider not storing the argument registers in the PROLOG_WITH_TRANSITION_BLOCK and
+        ; store / restore them only around this call. We would still need to save RCX and RDX as they can contain
+        ; the return buffer address.
+        call            CreateNativeToInterpreterCallStub
```

Comment:
I do not see how the arguments get reported to the GC while `CreateNativeToInterpreterCallStub` is executing. Is it a GC hole?

Would it be better to compile the transition helper in `PreStubWorker`? `PreStubWorker` gets called whenever we are about the execute a method from JIT/AOT code that is not compiled yet, and it has all the right infrastructure to protect arguments. It would be best to avoid duplicating it.

Reply:
That's a good point and compiling it in the `PreStubWorker` would definitely be better.